### ISSUE TITLE
DefaultIvyPublication no longer allows Usage other than Gradle compile/runtime in Gradle 4.1

### ISSUE THREAD

**Issue by Alex-Vol** [gradle/gradle#2865]
The DefaultIvyPublication uses Ordering.explicit(Usage.FOR_COMPILE, Usage.FOR_RUNTIME) to declare the ordering for UsageContext. This causes any Usage that is not of the two provided instances to be rejected. In previous versions of which I can attest only on version 2.10 Usage was not restricted to just the two predefined instances. In these versions the Usage was not compared which allowed values other than "compile" and "runtime" to coexist. I can see from the Github sources that this behavior is new in the 4.x series.

### Expected Behavior
I would like to suggest that this is changed to permit other usages to exist while still imposing the "compileBeforeRuntime" sorting when presented with these two specific Usage instances. To further reinforce this desire I submit the fact that in DefaultIvyPublication the private method `mapUsage(Usage)` does not preclude the possibility of other Usage instances.

### Current Behavior
Due to the implementation of the sorting currently this results in an exception when I provide my custom plugin SoftwareComponent as an argument to `from(SoftwareComponent)` method call.

```
Caused by: org.gradle.internal.impldep.com.google.common.collect.Ordering$IncomparableValueException: Cannot compare value: com.simplivity.gradle.plugins.nativeartifacts.NativeSoftwareComponent$NativeUsageContext$NativeUsage@3823c84a
	at org.gradle.internal.impldep.com.google.common.collect.ExplicitOrdering.rank(ExplicitOrdering.java:46)
	at org.gradle.internal.impldep.com.google.common.collect.ExplicitOrdering.compare(ExplicitOrdering.java:40)
	at org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublication$1.compare(DefaultIvyPublication.java:63)
	at org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublication$1.compare(DefaultIvyPublication.java:59)
	at java.util.TimSort.countRunAndMakeAscending(TimSort.java:351)
	at java.util.TimSort.sort(TimSort.java:216)
	at java.util.Arrays.sort(Arrays.java:1512)
	at java.util.ArrayList.sort(ArrayList.java:1454)
	at java.util.Collections.sort(Collections.java:175)
	at org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublication.getSortedUsageContexts(DefaultIvyPublication.java:145)
	at org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublication.from(DefaultIvyPublication.java:116)
```

If the implementation allowed a natural ordering of Usage.getName() unless the two very Maven specific Usages were presented it would work for my plugin implementation.

### Context
I am attempting to port a very functional and working plugin from Gradle 2.10 to Gradle 4.1. This plugin implements variant aware Publishing/Consuming of Native Artifacts from an Ivy repository. When I created this plugin a year and a half ago I intentionally chose Ivy to avoid the artificial Usage restrictions of Maven POM to have only compile/runtime usages. In doing so I accomplished a full featured solution that can correctly map dependencies in an easy to understand way between variants of buildType=debug/release, platform=linux/windows.

My solution, while specific to Ivy, is currently used in a large C++ native components project with 45 independently developed repositories contributing a collective of just as many shared libraries and executables. The CI system is capable of building these imposing semantic versioning rules on the dependencies and resolving artifacts according to build parameters. It allows developers to target specific build types and platforms or cross compiles in parallel all variants to publish via CI.

### Steps to Reproduce (for bugs)
Given a plugin that implements a new SoftwareComponentInternal type such as NativeSoftwareComponent with a NativeUsageContext and NativeUsage which implements Comparable I expect calling `IvyPublication.from(NativeSoftwareComponent)` to successfully configure my publication declaring the Ivy configurations based on the `Usage.getName()` results.

I can probably create a demonstration build.gradle with a little effort if necessary. But I think the issue is self evident in my opinion.

### Your Environment
Works using my plugin in Windows/Linux with Gradle 2.10. Fails while I am attempting to port my plugin to Gradle 4.1. I have not tried with Gradle 3.5.1 yet, I may attempt this if the `Ordering.explicit()` is not used in that version.

**Comment by Alex-Vol** [comment]
Although not immediately affecting my plugin operation the same issue seems to have come up in the DefaultMavenPublication class. The explicit ordering algorithm for compile/runtime prohibits the use of other custom Usage implementations.

In my mind Usage == Configuration and restricting this to compile/runtime is not going to get multi-variant builds where a compile-linux-debug artifact must depend on a compile-linux-debug dependency while a compile-window-release artifact must depend on a compile-windows-release dependency.

With my plugin implementation I have accomplished this level of multi-variant aware builds and is happily running in a Gradle 2.10 based CI system.

**Comment by bmuschko** [comment]
The classes `SoftwareComponentInternal`, `Usage` and `UsageContext` are all internal APIs. I'd avoid it if you can. You can follow a [similar discussion plus proposed alternative solutions here](https://github.com/gradle/gradle/issues/1361).

**Comment by Alex-Vol** [comment]
I am trying to solve a multiple variant problem that cannot be restricted in the compile/runtime dimensions alone as Java artifacts do. With native artifacts it is desirable to produce debug/release and windows/macos/linux variants from a single build.gradle configurations. With sufficient time and cross compilation toolchains it would even be possible to do this from a single build execution. In my working 2.10 Gradle implementation adding new Usage variants to capture the configuration names created by splicing together the variant combinations was producing fully declared dependency graphs in ivy.xml that could correctly resolve complex dependencies of the form linux-debug => linux-debug and windows-release => windows-release.

Removing the explicit ordering in the DefaultIvyPublication restored the ability to provide Usage instances with names other than the preconfigured compile/runtime set and allows the generated ivy.xml to be constructed with the appropriate dependency configurations as needed.

I cannot see a case where this can be made to work without this limitation relaxed. How can we produce variant aware software components without such a facility? I think the fact that it was fully functional even if tied to the SoftwareComponentInternal and Usage implementations proves that there is merit in the solution.

I will read on the discussion but I am not convinced alternatives are going to be sufficient.

**Comment by Alex-Vol** [comment]
I am also perplexed by the need to sort compile before runtime for the needs of publications. It seems arbitrary to enforce such a processing order in the code. What is the motivation of this sorting implementation? 

Allowing the natural ordering would accomplish the same just by coincidence but that is not relevant to the greater question about sorting.

**Comment by Alex-Vol** [comment]
I see the suggestion for the shadow plugin as a possible alternative to getting the ivy.xml generated instead of relying on the SoftwareComponent and the "IvyPublication.from(SoftwareComponent)". It puts the onus of constructing the ivy.xml entirely on my handler which I have done already for other reasons. 

That may be where I end up to solve the immediate problem but I have to add my vote for opening up the use cases for plugins contributing alternative SoftwareComponent implementations and letting the Usage be defined in an open public API instead of internalized for the exclusive use of Java and War plugins.

**Comment by oehme** [comment]
This has been fixed on master, please try the nightly build tomorrow. We needed a fix for this for our own native publishing efforts. Keep an eye on the release notes for that, you might not need your own plugin anymore after this is implemented. Opening up publishing for extension is another step we plan after that.

**Comment by Alex-Vol** [comment]
Thank you for the update @oehme. I have been monitoring the development of the native publishing with great anticipation. I can see that you are on the path of solving many of the requirements I have already in my own custom plugin. It will not be drop in compatible with our home grown plugin but as long as I can achieve parity with what we had I plan to completely replace the usage of our plugin with the standard Gradle plugins. 

I will try the latest nightly anyway because I had made progress in porting much of the plugin already aside from this blocking issue. I was able to create a custom private Gradle build with my own "fix" for this that proved I could still use my logic when Gradle did not cause this issue.
